File: src/main/java/com/skiwi/githubhooksechatservice/mvc/beans/tasks/CommentsScanTask.java
package com.skiwi.githubhooksechatservice.mvc.beans.tasks;

import java.time.Instant;
import java.util.Collections;
import java.util.List;

import org.apache.log4j.LogManager;
import org.apache.log4j.Logger;

import com.skiwi.githubhooksechatservice.chatbot.ChatBot;
import com.skiwi.githubhooksechatservice.init.CommentClassification;
import com.skiwi.githubhooksechatservice.mvc.beans.StackExchangeAPIBean;
import com.skiwi.githubhooksechatservice.mvc.controllers.WebhookParameters;
import com.skiwi.githubhooksechatservice.stackapi.StackComments;
import com.skiwi.githubhooksechatservice.stackapi.StackExchangeComment;

public class CommentsScanTask implements Runnable {
	private static final Logger logger = LogManager.getLogger(CommentsScanTask.class);
	
	private Instant nextFetch = Instant.now();
	private long lastComment;
	private long fromDate;
	private int remainingQuota;
	
	private final WebhookParameters params = WebhookParameters.toRoom("8595");
	private final WebhookParameters debug = WebhookParameters.toRoom("20298");
	private final WebhookParameters programmers = WebhookParameters.toRoom("21");
	private final WebhookParameters softwareRecs = WebhookParameters.toRoom("22668");

	private StackExchangeAPIBean stackAPI;

	private ChatBot chatBot;
	
	private Pattern interestingComment = Pattern.compile("[*`_]{,2}code[*`_]{,2}\s*[*`_]{,2}review[*`_]{,2}", Pattern.CASE_INSENSITIVE | Pattern.UNICODE_CASE);
	
	public CommentsScanTask(StackExchangeAPIBean stackAPI, ChatBot chatBot) {
		this.stackAPI = stackAPI;
		this.chatBot = chatBot;
	}

	private boolean isInterestingComment(StackExchangeComment comment) {
		String commentText = comment.getBodyMarkdown().toLowerCase();
		return interestingComment.matcher(commentText).find(0);
		// We're using `find()` instead of `matches()` because it can occur anywhere and we don't wanna reject if we can't match the whole comment
		// See https://docs.oracle.com/javase/8/docs/api/java/util/regex/Matcher.html#matches-- and #find-int- for more information
	}

	@Override
	public void run() {
		if (!Instant.now().isAfter(nextFetch)) {
			return;
		}

		try {
			StackComments comments = stackAPI.fetchComments("stackoverflow", fromDate);
			int currentQuota = comments.getQuotaRemaining();
			if (currentQuota > remainingQuota && fromDate != 0) {
				chatBot.postMessage(debug, Instant.now() + " Quota has been reset. Was " + remainingQuota + " is now " + currentQuota);
			}
			remainingQuota = currentQuota;
			List<StackExchangeComment> items = comments.getItems();
			if (items != null) {
				if (items.size() >= 100) {
					chatBot.postMessage(debug, Instant.now() + " Warning: Retrieved 100 comments. Might have missed some.");
				}
				
				long previousLastComment = lastComment;
				Collections.reverse(items);
				for (StackExchangeComment comment : items) {
					if (comment.getCommentId() <= previousLastComment) {
						continue;
					}
					lastComment = Math.max(comment.getCommentId(), lastComment);
					fromDate = Math.max(comment.getCreationDate(), fromDate);
					if (isInterestingComment(comment)) {
						chatBot.postMessage(params, comment.getLink());
					}
					float programmersCertainty = CommentClassification.calcInterestingLevelProgrammers(comment);
					
					if (programmersCertainty >= CommentClassification.REAL) {
						chatBot.postMessage(programmers, comment.getLink());
					}
					if (programmersCertainty >= CommentClassification.DEBUG) {
						chatBot.postMessage(debug, "Certainty level " + programmersCertainty);
						chatBot.postMessage(debug, comment.getLink());
					}
					
					float softwareCertainty = CommentClassification.calcInterestingLevelSoftwareRecs(comment);
					
					if (softwareCertainty >= CommentClassification.REAL) {
						chatBot.postMessage(softwareRecs, comment.getLink());
					}
				}
				items.clear();
			}
			if (comments.getBackoff() != 0) {
				nextFetch = Instant.now().plusSeconds(comments.getBackoff() + 10);
				chatBot.postMessage(debug, Instant.now() + " Next fetch: " + nextFetch + " because of backoff " + comments.getBackoff());
			}
		} catch (Exception e) {
			logger.error("Error retrieving comments", e);
			chatBot.postMessage(debug, Instant.now() + " Exception in comment task " + e);
			return;
		}
	}

}

File: duga-ktor/src/utils/stackexchange/CommentsScanTask.kt
package net.zomis.duga.utils.stackexchange

import com.fasterxml.jackson.databind.JsonNode
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch

import java.time.Instant;

import net.zomis.duga.chat.DugaPoster
import net.zomis.machlearn.text.TextClassification
import org.slf4j.LoggerFactory

class CommentsScanTask(
	private val scope: CoroutineScope,
	private val stackAPI: StackExchangeApi,
	private val programmersClassification: TextClassification,
	poster: DugaPoster
) {
    private val logger = LoggerFactory.getLogger(CommentsScanTask::class.java)
    
    private var nextFetch = Instant.now()
    private var lastComment: Long = 0
    private var fromDate: Long = 0
    private var remainingQuota: Long = 0
    
	private val codeReview = poster.room("8595")
	private val debug = poster.room("20298")
	private val programmers = poster.room("21")
	private val softwareRecs = poster.room("20298")

	suspend fun run() {
		val now = Instant.now()
    	if (!now.isAfter(nextFetch)) {
    		logger.info("Comments scan next fetch is at $nextFetch and now is $now, need to wait a bit more, returning.")
    		return;
    	}

		val comments = stackAPI.fetchComments("stackoverflow", fromDate)
		if (comments == null) {
			logger.error("Unable to get comments from $fromDate")
			return
		}
		val currentQuota = comments.get("quota_remaining").asLong()
		if (currentQuota > remainingQuota && fromDate != 0L) {
			debug.post(Instant.now().toString() + " Quota has been reset. Was " +
				remainingQuota + " is now " + currentQuota)
		}
		remainingQuota = currentQuota

		if (comments.get("backoff") != null) {
			nextFetch = Instant.now().plusSeconds(comments.get("backoff").asLong() + 10)
			debug.post(Instant.now().toString() +
					" Next fetch: " + nextFetch + " because of backoff " + comments.get("backoff"))
			return
		}

		val items = comments.get("items") ?: return
		if (items.size() >= 100) {
			debug.post(Instant.now().toString() + " Warning: Retrieved 100 comments. Might have missed some.")
		}

		val previousLastComment = lastComment
		for (comment in items.reversed()) {
			if (comment.get("comment_id").asLong() <= previousLastComment) {
				continue;
			}
			lastComment = Math.max(comment.get("comment_id").asLong(), lastComment)
			fromDate = Math.max(comment.get("creation_date").asLong(), fromDate)
			if (isInterestingCommentCR(comment)) {
				logComment(comment, "Code Review")
				codeReview.post(comment.get("link").asText())
			}

			classifyProgrammers(comment)

			val softwareCertainty = CommentClassification.calcInterestingLevelSoftwareRecs(comment)

			if (softwareCertainty >= CommentClassification.REAL) {
				softwareRecs.post(comment.get("link").asText())
			}
		}
    }

	private fun logComment(comment: JsonNode, site: String) {
		logger.info("$site comment $comment.comment_id " +
				"on $comment.post_type $comment.post_id " +
				"posted by $comment.owner.display_name " +
				"with $comment.owner.reputation reputation: ${comment.get("body_markdown")}")
	}

	private fun isInterestingCommentCR(comment: JsonNode): Boolean {
		val commentText = comment.get("body_markdown").asText().toLowerCase()
		return commentText.contains("code review") || commentText.contains("codereview")
	}

	private fun classifyProgrammers(comment: JsonNode) {
        val oldClassification = CommentClassification.calcInterestingLevelProgrammers(comment);
        val programmersMLscore = programmersMLscore(comment)

        if (programmersMLscore >= CommentClassification.ML_THRESHOLD) {
            programmers.postAsync(scope, comment.get("link").asText())
        }

        if (programmersMLscore >= CommentClassification.DEBUG) {
			logComment(comment, "Software Engineering (ML $programmersMLscore old $oldClassification)")
            val certaintyLevelMessage =
                    "ML Classification " + programmersMLscore +
                            " (Old classification " + oldClassification + ")";
			scope.launch {
				debug.post(certaintyLevelMessage)
				debug.post(comment.get("link").asText())
			}
        }
    }

    private fun programmersMLscore(comment: JsonNode): Double {
        val text = comment.get("body_markdown").asText().toLowerCase()
        if (!text.contains("programmers") && !text.contains("softwareeng")
	     && !text.contains("software eng")) {
            // No need to check with the Machine Learning system in this case
            return -1.0
        }
        return programmersClassification.score(text)
    }
}
